File: app/src/main/java/com/repelliuss/npuzzle/game/NPuzzle.java
package com.repelliuss.npuzzle.game;

import com.repelliuss.npuzzle.utilities.Index2D;

import java.lang.reflect.Array;

public final class NPuzzle extends SlidePuzzle<Integer> {

    private final Piece[] board;
    private int row, column;
    private final Index2D posBlank;

    public NPuzzle(int argRow, int argColumn) {
        row = argRow;
        column = argColumn;

        //noinspection unchecked
        board = (Piece[]) Array.newInstance(Piece.class, row * column);

        for(int i = 0; i < row * column - 1; ++i) board[i] = new Piece(Cell.VALUE, i + 1);
        board[row * column - 1] = new Piece(Cell.BLANK);
        posBlank = new Index2D(row - 1, column - 1);
    }

    @Override
    public int getRow() { return row; }

    @Override
    public int getColumn() { return column; }

    @Override
    public Piece getPiece(Index2D pos) {
        return board[getIndex(pos)];
    }

    @Override
    protected Index2D getPosBlank() { return posBlank; }

    @Override
    protected void setRow(int argRow) { row = argRow; }

    @Override
    protected void setColumn(int argColumn) { column = argColumn; }

    @Override
    protected void setPiece(Index2D pos, final Piece piece) {
        board[getIndex(pos)].setPiece(piece);
    }

    @Override
    public void reset() {
        int number = 1;
        Index2D pos = new Index2D();

        for(; pos.getY() < getRow(); pos.incY()) {
            for(pos.setX(0); pos.getX() < getColumn() - 1; pos.incX()) {
                getPiece(pos).setPiece(Cell.VALUE, number++);
            }
            number += 2;
        }

        number = getColumn();

        for(pos.setTo(0, getColumn() - 1); pos.getY() < getRow() - 1; pos.incY()) {
            getPiece(pos).setPiece(Cell.VALUE, number);
            number += getColumn();
        }

        pos.setTo(getRow() - 1, getColumn() - 1);
        getPiece(pos).setId(Cell.BLANK);
    }

    @Override
    public boolean isSolved() {

        Index2D pos = new Index2D(getRow() - 1, getColumn() - 1);

        if(getPiece(pos).getId() != Cell.BLANK)
            return false;

        int number = 1;

        for (pos.setY(0); pos.getY() < getRow(); pos.incY()) {
            for (pos.setX(0); pos.getX() < getColumn() - 1; pos.incX()) {
                if (getPiece(pos).getValue() != number++) return false;
            }
            number += 2;
        }

        number = getColumn();

        for (pos.setTo(0, getColumn() - 1); pos.getY() < getRow() - 1; pos.incY()) {
            if (getPiece(pos).getValue() != number) return false;
            number += getColumn();
        }

        return true;
    }

    private int getIndex(final Index2D index2D) {
        return getColumn() * index2D.getY() + index2D.getX();
    }
}

File: deep_learning/game/npuzzle.py
#!/usr/bin/env python3

from game.slidepuzzle import SlidePuzzle
from game.piece import Piece
from game.cell import Cell
from game.move import Move
import numpy as np

class NPuzzle(SlidePuzzle):

    def __init__(self, row, column):
        super().__init__()
        self.__row = row
        self.__column = column
        self.__board = [[Piece(Cell.VALUE, y*column + x + 1) for x in range(column)]
                        for y in range(row)]
        self.__board[-1][-1] = Piece(Cell.BLANK, 0)
        self.__pos_blank = (row-1, column-1)

    def p(self):
        for y in range(self.__row):
            for x in range(self.__column):
                print(self.__board[y][x].value)

    @property
    def row(self):
        return self.__row

    @row.setter
    def _SlidePuzzle__row(self, row):
        self.__row = row

    @property
    def column(self):
        return self.__column

    @column.setter
    def _SlidePuzzle__column(self, column):
        self.__column = column

    def get_pos_blank(self):
        return self.__pos_blank

    def _SlidePuzzle__update_pos_blank(self, delta_y, delta_x):
        self.__pos_blank = (self.__pos_blank[0] + delta_y,
                            self.__pos_blank[1] + delta_x)

    def get_cell(self, pos):
        return self.__board[pos[0]][pos[1]]

    def _SlidePuzzle__set_cell(self, pos, other):
        self.get_cell(pos).copy_piece(other)

    def reset(self):
        number = 1
        for y in range(self.row):
            for x in range(self.column):
                self.get_cell((y,x)).set_piece(Cell.VALUE, number)
                number += 1

        self.get_cell((self.row - 1, self.column - 1)).set_piece(Cell.BLANK, 0)
        self.__pos_blank = (self.row - 1, self.column - 1)


    def is_solved(self):
        if self.get_cell(self.get_pos_blank()).id != Cell.BLANK:
            return False

        number = 1
        for y in range(self.row):
            for x in range(self.column - 1):
                if self.get_cell((y,x)).value != number:
                    return False
                number += 1
            number += 1

        number = self.column
        for y in range(self.row - 1):
            if self.get_cell((y, self.column - 1)).value != number:
                return False
            number += self.column

        return True

    def numpy(self):
        matrix = np.zeros((self.row, self.column))
        for i in range(self.row):
            for j in range(self.column):
                matrix[i][j] = self.get_cell((i,j)).value
        return matrix
